modules/tinymce/src/core/test/ts/browser/UndoManagerTest.ts

```diff
@@ -611,4 +612,68 @@ describe('browser.tinymce.core.UndoManagerTest', () => {
 
     assert.equal(changeEventCounter, 0, 'No events should be detected');
   });
+
+  it('TINY-8641: Dispatch change and set editor dirty if the current content and the undoManager history are incoherent', () => {
+    const editor = hook.editor();
+
+    let changeEventCounter = 0;
+    let currentChangeEvent: { level: string; lastLevel: string };
+
+    const initialContent = '<p>some inital content</p>';
+    const manualModifiedLevel = 'a modified last level';
+
+    const assertChangeEvent = (event: { level: string; lastLevel: string }) => {
+      assert.equal(event.level, initialContent, 'Level has not the expected content');
+      assert.equal(event.lastLevel, manualModifiedLevel, 'Last level has not the expected content');
+    };
+
+    editor.resetContent(initialContent);
+    const onChange = (e: EditorEvent<{
+      level: UndoLevel;
+      lastLevel: UndoLevel;
+    }>) => {
+      changeEventCounter++;
+      currentChangeEvent = { level: e.level.content, lastLevel: e.lastLevel.content };
+    };
+
+    editor.on('change', onChange);
+
+    assert.equal(changeEventCounter, 0, '0 event should be detected at start');
+    editor.undoManager.dispatchIfChanged();
+    assert.equal(changeEventCounter, 0, '0 event should be detected if content is coherent with the undoManager history');
+
+    Arr.last(editor.undoManager.data).each((lastLevel) => {
+      lastLevel.content = manualModifiedLevel;
+    });
+
+    assert.equal(editor.isDirty(), false, 'Editor should not be dirty before dispatchIfChanged');
+    editor.undoManager.dispatchIfChanged();
+    assertChangeEvent(currentChangeEvent);
+    assert.equal(changeEventCounter, 1, '1 event should be detected if the content is not coherent with the undoManager history');
+    assert.equal(editor.isDirty(), true, 'Editor should be dirty after dispatchIfChanged');
+
+    editor.undoManager.dispatchIfChanged();
+    assertChangeEvent(currentChangeEvent);
+
+    editor.undoManager.dispatchIfChanged();
+    assertChangeEvent(currentChangeEvent);
+
+    editor.undoManager.dispatchIfChanged();
+    assertChangeEvent(currentChangeEvent);
+    assert.equal(changeEventCounter, 4, 'it should continue to call change till the editor and last level are different');
+
+    editor.undoManager.add();
+    assert.equal(changeEventCounter, 5, 'add should trigger a change as well');
+
+    Arr.last(editor.undoManager.data).each((lastLevel) => {
+      assert.equal(editor.getContent(), lastLevel.content, 'add should add a last layer with the editor content');
+    });
```

Comment:
This assertion isn't going to be working as you might expect... If there is no last level then it's not going to fail. Generally in tests you want to assert the length of the array as well before accessing the item.

- [ ] Improve this assertion to fail if there is no `data`

Reply:
done